### PR TITLE
feat: use molecule to serialize witness

### PR DESCRIPTION
**Pull request by jjyr**
Changes:

* The witnesses struct is changed from raw bytes to WitnessArgs , both for `secp256k1_blake160_sighash_all` and `secp256k1_ripemd160_sha256_sighash_all`.
* The `sighash` behavior of `secp256k1_ripemd160_sha256_sighash_all` is changed,  the lock will digest other witnesses into `sighash` to protecting them.
* Fix security issue: witnesses ambiguity https://github.com/nervosnetwork/ckb-system-scripts/pull/54/files#diff-74a4a9fca557c2531a16e101766eff1cR508
* Deny txs which witnesses are not equal to inputs.

The cycles changed from `3390910` to `3406275` for a 2-in-2-out tx, but the main reason is that the `signature` changed, cause secp256k1 consumed more(or less) cycles.

### CODE REVIEW

**Comment by janx** [comment]
The extra cost is quite small.

File: c/common.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#define ERROR_SECP_PARSE_PUBKEY -13
+#define ERROR_SECP_PARSE_SIGNATURE -14
+#define ERROR_SECP_SERIALIZE_PUBKEY -15
+#define ERROR_SCRIPT_TOO_LONG -21
+#define ERROR_WITNESS_SIZE -22
+#define ERROR_INVALID_WITNESSES_COUNT -23
+#define ERROR_PUBKEY_BLAKE160_HASH -31
+#define ERROR_PUBKEY_RIPEMD160_HASH -32
+
+/* make sure witnesses is equals to inputs */
```
Comment (doitian) [inline c/common.h]:
```suggestion
/* make sure that witnesses and inputs have the same length */
```

File: c/common.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+int check_witnesses_len() {
+  uint64_t len = 0;
+  uint8_t tmp[0];
+  /* lower bound, at least tx has one input */
+  int lo = 0;
+  /* higher bound */
+  int hi = 4;
+  int ret;
+  /* try loading input until failed to increase lo and hi */
```
Comment (doitian) [inline c/common.h]:
```suggestion
  /* try to load input until failing to increase lo and hi */
```

File: c/common.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    i = (lo + hi) / 2;
+    ret = ckb_load_input_by_field(tmp, &len, 0, i, CKB_SOURCE_INPUT,
+                                  CKB_INPUT_FIELD_SINCE);
+    if (ret == CKB_SUCCESS) {
+      lo = i;
+    } else {
+      hi = i;
+    }
+  }
+  /* count of witnesses should equals to inputs */
```
Comment (doitian) [inline c/common.h]:
```suggestion
  /* witnesses and inputs must have the same length */
```

File: c/secp256k1_blake160_sighash_all.c [hunk 4]
```diff
@@ -1,191 +1,180 @@
...
 
+  /* Prepare sign message */
+  unsigned char message[BLAKE2B_BLOCK_SIZE];
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+
+  /* Clear lock field to zero, then digest the first witness */
+  memset((void *)lock_bytes_res.pos.ptr, 0, lock_bytes_res.pos.size);
+  blake2b_update(&blake2b_ctx, (char *)&witness_len, sizeof(uint64_t));
```
Comment (yangby-cryptape) [inline c/secp256k1_blake160_sighash_all.c]: I wish 3rd-party developers can understand this: `(char *)&witness_len` (since little endian).

**Pull request merged**